nhrp: add `cisco-authentication` password support

volodymyrhuti opened the pull request:
Implemented:
- handling 8 char long password, aka Cisco style.
- minimal error inidication routine
- test case, password change affects connection

Implemented in the context - https://vyos.dev/T2326

ton31337 commented on nhrpd/nhrp_vty.c:
```diff
@@ -459,6 +460,54 @@ DEFUN(if_no_nhrp_holdtime, if_no_nhrp_holdtime_cmd,
 }
 
+#define NHRP_CISCO_PASS_LEN 8
+DEFUN(if_nhrp_authentication, if_nhrp_authentication_cmd,
```
DEFPY please

volodymyrhuti replied:
fixed

ton31337 commented on nhrpd/nhrp_vty.c:
```diff
@@ -459,6 +460,54 @@ DEFUN(if_no_nhrp_holdtime, if_no_nhrp_holdtime_cmd,
 
+#define NHRP_CISCO_PASS_LEN 8
+DEFUN(if_nhrp_authentication, if_nhrp_authentication_cmd,
+      "nhrp authentication PASS",
```
PASS -> WORD, please

volodymyrhuti replied:
fixed

ton31337 commented on nhrpd/nhrp_vty.c:
```diff
@@ -459,6 +460,54 @@ DEFUN(if_no_nhrp_holdtime, if_no_nhrp_holdtime_cmd,
+	int pass_len = strlen(pass);
+
+	if (pass_len > NHRP_CISCO_PASS_LEN) {
+		vty_out(vty, "Password size limit exceeded (%d>%d)", pass_len,
```
new line is missing for vty_out()'s output.

volodymyrhuti replied:
fixed

ton31337 commented on nhrpd/nhrp_vty.c:
```diff
@@ -459,6 +460,54 @@ DEFUN(if_no_nhrp_holdtime, if_no_nhrp_holdtime_cmd,
+}
+
+
+DEFUN(if_no_nhrp_authentication, if_no_nhrp_authentication_cmd,
```
DEFPY

volodymyrhuti replied:
fixed

ton31337 commented on nhrpd/nhrp_vty.c:
```diff
@@ -459,6 +460,54 @@ DEFUN(if_no_nhrp_holdtime, if_no_nhrp_holdtime_cmd,
+
+
+DEFUN(if_no_nhrp_authentication, if_no_nhrp_authentication_cmd,
+      "no nhrp authentication",
```
We should allow an operator to enter `no nhrp authentication BLABLA` too.

ton31337 commented on nhrpd/nhrp_vty.c:
```diff
@@ -459,6 +460,54 @@ DEFUN(if_no_nhrp_holdtime, if_no_nhrp_holdtime_cmd,
+{
+	VTY_DECLVAR_CONTEXT(interface, ifp);
+	struct nhrp_interface *nifp = ifp->info;
+	zbuf_free(nifp->auth_token);
```
Shouldn't we check against NULL?

volodymyrhuti replied:
yeap, fixed

ton31337 commented on nhrpd/nhrp_vty.c:
```diff
@@ -1155,6 +1205,12 @@ static int interface_config_write(struct vty *vty)
 			vty_out(vty, " tunnel source %s\n", nifp->source);
 
+		if (nifp->auth_token) {
+			auth = (struct nhrp_cisco_authentication_extension *)
```
nit: typecasting is not needed (same type as left operand).

volodymyrhuti replied:
doesn't seem to be the case
auth             | nhrp_cisco_authentication_extension
auth_token | zbuf*
buf               | uint8_t*

ton31337 reviewed (changes requested):
Missing documentation, some contributing guidelines, and styling issues yet.

pguibert6WIND commented on nhrpd/nhrp_packet.c:
```diff
@@ -115,14 +115,35 @@ uint16_t nhrp_packet_calculate_checksum(const uint8_t *pdu, uint16_t len)
-void nhrp_packet_complete(struct zbuf *zb, struct nhrp_packet_header *hdr)
+void nhrp_packet_complete(struct zbuf *zb, struct nhrp_packet_header *hdr,
+			  struct interface *ifp) {
+	nhrp_packet_complete_auth(zb, hdr, ifp, true)
```
missing ; in the end.

volodymyrhuti replied:
oops, my bad, fixed

pguibert6WIND commented:
general remark: apply 'git clang-format HEAD^' and squash the proposed changes.

donaldsharp commented:
@louberger can you take a look at this?

volodymyrhuti commented:
The latest update:
- addressed the PR comments
- fixed styling with clang-format 
- added a short doc for the password config

The internal testing by my team has shown issues when connecting with Cisco Spoke; keep PR as a draft until addressed

fett0 commented:
@ton31337  @pguibert6WIND    is anything else needed to tmerge this PR ?

dleroy commented:
@volodymyrhuti I also have an interest in seeing this committed. I tested your PR against a Cisco router as spoke and found/fixed a few issues. (see attached file).  Wireshark is still complaining. I'm wondering if its because the packet causing the error indication does not include the extensions when copying the packet that caused the indication. Wireshark just recursively parses the included original packet, so the checksum and length would be incorrect.

Issues I found/fixed:
1) FRR would accept messages from a spoke without authentication when FRR NHRP had auth configured.
2) The error indication was not being sent in network byte order
3) the debug print in nhrp_connection_authorized was not correctly printing the received password
4) the addresses portion of the mandatory part of the error indication were invalid on the wire (confirmed in wireshark)

Happy to discuss, let me know if you have any questions.

[auth.txt](https://github.com/FRRouting/frr/files/15420658/auth.txt)

Jafaral commented:
Replaced by https://github.com/FRRouting/frr/pull/16172
